tests/testapp/tests/test_connection.py

```diff
@@ -0,0 +1,134 @@
+            scope_definition=self.subset_scope_def,
+            scope_version=self.subset_scope_def.version,
+            scope_params=json.dumps({"mainpartition": self.root_cert.id, "subpartition": "abracadabra"}),
+            private_key=Key(),
+        )
+        self.root_cert.sign_certificate(self.subset_cert)
+        self.subset_cert.save()
+
+        self.controller = MorangoProfileController('facilitydata')
+        self.network_connection = self.controller.create_network_connection('127.0.0.1')
+
+    @mock_patch_decorator
+    def test_creating_sync_session_successful(self):
+        self.assertEqual(SyncSession.objects.filter(active=True).count(), 0)
+        self.network_connection.create_sync_session(self.subset_cert, self.root_cert)
+        self.assertEqual(SyncSession.objects.filter(active=True).count(), 1)
+
+    @mock_patch_decorator
+    def test_creating_sync_session_cert_fails_to_verify(self):
+        Certificate.verify.return_value = False
+        with self.assertRaises(CertificateSignatureInvalid):
+            self.network_connection.create_sync_session(self.subset_cert, self.root_cert)
+
+    @mock_patch_decorator
+    def test_get_remote_certs(self):
+        # mock certs being returned by server
+        certs = self.subset_cert.get_ancestors(include_self=True)
+        cert_serialized = json.dumps(CertificateSerializer(certs, many=True).data)
+        Connection._request.return_value = mock.Mock(data=cert_serialized)
+
+        # we want to see if the models are created (not saved) successfully
+        remote_certs = self.network_connection.get_remote_certificates('abc')
+        self.assertSetEqual(set(certs), set(remote_certs))
+
+    @mock_patch_decorator
+    def test_csr(self):
+        # mock a "signed" cert being returned by server
+        cert_serialized = json.dumps(CertificateSerializer(self.subset_cert).data)
+        Connection._request.return_value = mock.Mock(data=cert_serialized)
+        self.subset_cert.delete()
+
+        # we only want to make sure the "signed" cert is saved
+        with mock.patch.object(Key, "get_private_key_string", return_value=self.subset_cert.private_key.get_private_key_string()):
+            self.network_connection.certificate_signing_request(self.root_cert, '', '')
+        self.assertTrue(Certificate.objects.filter(id=json.loads(cert_serialized)['id']).exists())
+
+    @mock_patch_decorator
+    def test_get_cert_chain(self):
+        # mock a cert chain being returned by server
+        certs = self.subset_cert.get_ancestors(include_self=True)
+        cert_serialized = json.dumps(CertificateSerializer(certs, many=True).data)
+        Connection._request.return_value = mock.Mock(data=cert_serialized)
+        Certificate.objects.all().delete()
+
+        # we only want to make sure the cert chain is saved
+        self.network_connection._get_certificate_chain(certs[1])
+        self.assertEqual(Certificate.objects.count(), certs.count())
```

Comment:
This may be trivially true; would be best to evaluate `certs.count()` into a var at the top, rather than after stuff has been deleted.

Reply:
Done!